copulas/multivariate/Tree.py

```diff
@@ -0,0 +1,379 @@
+        return tau
+
+    def _data4next_T(self):
+        """
+        prepare conditional U matrix for next tree
+        """
+        # U = np.empty([self.n_nodes,self.n_nodes],dtype=object)
+        edge_set = self.edge_set
+        for k in range(len(edge_set)):
+            edge = edge_set[k]
+            copula_name = c_map[edge.name]
+            copula_para = edge.param
+            if self.level == 1:
+                U1, U2 = self.u_matrix[:, edge.L], self.u_matrix[:, edge.R]
+            else:
+                prev_T = self.prev_T.edge_set
+                l_p = edge.parent[0]
+                r_p = edge.parent[1]
+                U1, U2 = prev_T[l_p].U, prev_T[r_p].U
+            # compute conditional cdfs C(i|j) = dC(i,j)/duj and dC(i,j)/dui'''
+            U1 = [x for x in U1 if x is not None]
+            U2 = [x for x in U2 if x is not None]
+            c1 = Copula(cname=copula_name)
+            c1.fit(U1, U2)
+            derivative = c1.get_h_function()
+            U1givenU2 = derivative(U2, U1, copula_para)
+            U2givenU1 = derivative(U1, U2, copula_para)
+            # correction of 0 or 1'''
+            U1givenU2[U1givenU2 == 0], U2givenU1[U2givenU1 == 0] = eps, eps
+            U1givenU2[U1givenU2 == 1], U2givenU1[U2givenU1 == 1] = 1 - eps, 1 - eps
+            edge.U = U1givenU2
+
+    def _likehood_T(self, U):
+        """Compute likelihood of the tree given an U matrix
+        """
+        newU = np.empty([self.vine.n_var, self.vine.n_var])
+        edge_set = self.edge_set
+        values = np.zeros([1, len(edge_set)])
+        for i in range(len(edge_set)):
+            edge = edge_set[i]
+            cname = self.vine.c_map[edge.name]
+            v1 = edge.L
+            v2 = edge.R
+            copula_para = edge.param
+            if self.level == 1:
+                U1 = np.array([U[v1]])
+                U2 = np.array([U[v2]])
+            else:
+                prev_T = self.prev_T.edge_set
+                l_p = edge.parent[0]
+                r_p = edge.parent[1]
+                U1, U2 = prev_T[l_p].U, prev_T[r_p].U
+            cop = Copula(cname=cname)
+            cop.fit(U1, U2)
+            pdf = cop.get_pdf()
+            derivative = cop.get_h_function()
+            values[0, i] = pdf(U1, U2)
+            U1givenU2 = derivative(U2, U1, copula_para)
+            # U2givenU1 = derivative(U1, U2, copula_para)
+            edge.U = U1givenU2
+            value = np.sum(np.log(values))
+        return newU, value
+
+    def print_tree(self):
+        for e in self.edge_set:
+            msg = "L:{} R:{} D:{} parent:{}".format(e.L, e.R, e.D, e.parent)
+            LOGGER.info(msg)
+
+
+class CTree(Tree):
+    def __init__(self, k, n, tau_mat, prev_T):
+        super(CTree, self).__init__(k, n, tau_mat, prev_T)
+
+    def _build_first_tree(self):
+        # first column is the variable of interest
+        tau_mat = self.tau_mat
```

Comment:
As commented before, this line is unnecesary, you can access `self.tau_mat` directly